File: cPlusPlusCode/camScannerPerspectiveCorrection.cpp
#include<opencv2/imgproc.hpp>
#include<opencv2/highgui.hpp>
#include<opencv2/calib3d.hpp>
#include<iostream>

using namespace std;
using namespace cv;

int main(){
    //Read the book image
    Mat book = imread("../assets/book.jpg");

    vector<Point2f> pts_book, pts_correct;

    //four corners of the book in the image
    pts_book.push_back(Point2f(1456, 1580));
    pts_book.push_back(Point2f(2237, 2048));
    pts_book.push_back(Point2f(1417, 3668));
    pts_book.push_back(Point2f(604, 2744));

    //general aspect ratio of books are 3:4 for so we choose the size (300, 400)
    pts_correct.push_back(Point2f(0, 0));
    pts_correct.push_back(Point2f(299, 0));
    pts_correct.push_back(Point2f(299, 399));
    pts_correct.push_back(Point2f(0, 399));

    //calculate homography from book image to size(300, 400)
    Mat homographyMat = findHomography(pts_book, pts_correct);

    Mat result;

    //warp book for perspective correction (300, 400)
    warpPerspective(book, result, homographyMat, Size(300, 400));

    //create windows to display images
    namedWindow("book", WINDOW_NORMAL);
    namedWindow("correction", WINDOW_NORMAL);

    //display image
    imshow("book", book);
    imshow("correction", result);

    //Press esc to exit the program
    waitKey(0);

    //Close all the opened windows
    destroyAllWindows();
    
    return 0;
}

File: pythonCode/camScannerPerspectiveCorrection.py
import cv2
import numpy as np

#Read the book image
book  = cv2.imread("../assets/book.jpg")

#four corners of the book in the image
pts_book = np.array([[1456, 1580], [2237, 2048], [1417, 3668], [604, 2774]])

#general aspect ratio of books are 3:4 for so we choose the size (300, 400)
pts_correct = np.array([[0, 0], [299, 0], [299, 399], [0, 399]])

#calculate homography from book image to size(300, 400)
homorgaphyMat, status = cv2.findHomography(pts_book, pts_correct)

#warp book for perspective correction (300, 400)
result = cv2.warpPerspective(book, homorgaphyMat, (300, 400))

#create windows to display images
cv2.namedWindow("book", cv2.WINDOW_NORMAL)
cv2.namedWindow("correction", cv2.WINDOW_NORMAL)

#display images
cv2.imshow("book", book)
cv2.imshow("correction", result)

#press esc to exit the program
cv2.waitKey(0)

#close all the opened windows
cv2.destroyAllWindows()